### PR TITLE
Version 2.0

### LINKED ISSUE #112: React Router v4 Documentation & Examples
With the removal of `onEnter` and `onChange` in React Router v4 I think the auth-wrapper library makes even more sense. The docs talk about using the component lifecycle in your components for these features which is what redux-auth-wrapper already does!

Should be pretty easy just examples & docs, maybe a small breaking change in how the location/redirect is passed down

https://github.com/ReactTraining/react-router/blob/v4/README.md

### LINKED ISSUE #127: Router redirect feature incompatible with React Router v4.0.0-beta.7
The latest release of React Router causes the 'use router for redirect if no redirectAction specified' feature to break.

From the [React Router release notes](https://github.com/ReactTraining/react-router/releases/tag/v4.0.0-beta.7)

> Split `context.router` into two parts: `context.history` and `context.route`

This causes [L126](https://github.com/mjrussell/redux-auth-wrapper/blob/4e0145bf55cabdd4597afa41241fd068ef640c99/src/index.js#L126) and [L130](https://github.com/mjrussell/redux-auth-wrapper/blob/4e0145bf55cabdd4597afa41241fd068ef640c99/src/index.js#L130) to error when using `redux-auth-wrapper` without a `redirectAction` and with `react-router v4.0.0-beta.7`

### LINKED ISSUE #136: URL Module error on React Native
Hello,

I added redux-auth-wrapper and tried to run it in the code but it says following error:
UnableToResolveError: Unable to resolve module `url` from `.../node_modules/redux-auth-wrapper/lib/index.js`. I checked and there is no url module in node_modules.

I'm using react-native with react-native-router-flux. What should I do?

Thanks!

### LINKED ISSUE #140: Optimize options to be isolated with react-router or more friendly with other routers
Now, I succeed to use this repo without react-router. But it's kind of a [trial](https://github.com/mjrussell/redux-auth-wrapper/pull/139).
I've proposed [PR](https://github.com/mjrussell/redux-auth-wrapper/pull/139) with docs for other routers, so I think it means `redux-auth-wrapper` should change a little bit to be itself but not `react-router-redux-auth-wrapper`

### PR DESCRIPTION
**Pull request by mjrussell**
Its almost here! 🎉 

Lots of work to get to this point, but I believe the API to be stable, and I've got all the tests passing with full coverage. Although this is a complete rewrite internally and I've made it much more flexible to interact with React Native / other routers, the API is very similar and should be familiar to most users. There are a few breaking changes mainly to cleanup poor naming and only one feature removal (`propMapper`) because it made it easy to accidentally break the redirection. However that requirement can easily be solved by mapping props at the `AuthenticatingComponent`, `FailureComponent`, or the wrapped component.

The main outstanding work is in the documentation, I'm going to work on porting it to gitbook in order to be easier to read and navigate.

For some background, see #145 

Resolves #112
Resolves #127
Resolves #136
Resolves #140

### CODE REVIEW

**Comment by pepesenaris** [comment]
@mjrussell I'm starting a new project with react-router v4. If I understand correctly, from the issues and pull requests cited here, redux-auth-wrapper will not compatible with RR v4 until this PR is merged. is that right?
Cheers and thanks for the good work!!!

**Comment by mgambati** [comment]
When will the branch be merged? I'm starting a react-router v4 project too. 
With full test coverage the only thing that is needed is new docs?

**Comment by mjrussell** [comment]
@pepesenaris @mgambati 

Sorry for the delay, having trouble finding time to put the finishing touches on this. 

If you are using React Router v4 yes I would recommend using the new version (`npm install redux-auth-wrapper@2.0-beta`). Im considering a small-ish breaking change that removes the `authSelector` and `predicate` and combines them into a single `authenticatedSelector`. Previously it was useful to split them and pass down the result of `authDataSelector` since react-redux's connect couldn't guarantee that children would be updated from the store after their parent. Now that its been fixed, you can easily connect a child to grab the `authData` if you have a need for it.

**Comment by pepesenaris** [comment]
@mjrussell, thanks man. I'll go for the beta then.

**Comment by jwickens** [comment]
I am successfully using the beta in my app thanks @mjrussell !

Only changes necessary from the router v4 example were:
1) 
```
import locationHelperBuilder from 'redux-auth-wrapper/lib/history4/locationHelper'
import { connectedRouterRedirect } from 'redux-auth-wrapper/lib/history4/redirect'
```
instead of 
```
import locationHelperBuilder from 'redux-auth-wrapper/history4/locationHelper'
import { connectedRouterRedirect } from 'redux-auth-wrapper/history4/redirect'
```

and locationHelper has `locationHelper.getRedirectQuery(ownProps) ` rather than `locationHelper.getRedirectQueryParams(ownProps) `

**Comment by mjrussell** [comment]
@jwickens glad it worked for you. Good catch on 2, I'll fix that up.

I'll look into 1 also, it might be a little tricky due to how I'm using webpack's alias to depend on the source directly. Not sure if I can use the `lib` there like how it will be imported in a real world example. If I cant I'll at least put a comment above the import line. Thanks!

Going to put some hours into this tomorrow, I know its long overdue

**Comment by rolele** [comment]
Can I use the rewrite-v2 branch in my project with rr4 already?
Is this branch kind of stable? any plan on merging soon?

Great work! Thank you very much.
I think there is no other good auth library for v4 (they all look kind of unmaintained).
I am wondering how is this possible. Is the redux trend kind of slowing down?

**Comment by piuccio** [comment]
I've been using the beta2 branch with react-router 4 for a while. Seems ready to be shipped, there's only one thing that trips me up everytime but I'm not sure it's related to v2:

```
export const withAuth = connectedRouterRedirect({
  wrapperDisplayName: 'Authenticated',
  authSelector: makeSelectAuthData(),
  authenticatingSelector: () => true,
  AuthenticatingComponent: Loading,
  redirectPath: '/', // this should be ignored, I'm using the FailureComponent
  predicate: () => true,
  FailureComponent: Login,
});
```

|  | authenticatingSelector: *true* | authenticatingSelector: *false* |
|-------------------------------------|------|-------|
| predicate: *true*                                |   **allow access**   |    Login   |
| predicate: *false*                               |   Loading   |   allow access    |

When the `authenticatingSelector` returns `true` I'd expect the `Loading` component to be rendered regardless of the result of the predicate. Is this the intended behaviour?

File: docs/Migrating.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Migrating from Version 1.x to Version 2.x
+
+## Motivation for 2.x
+
+redux-auth-wrapper has really changed a lot since it was first designed as a small specialized utility for handling redirection with react-router-redux. Since then, it has been adopted by many developers and they've used it in numerous unforeseen and useful ways such as hiding components, displaying alternate components, and integrating with other routers. Additionally, redux-auth-wrapper needed to support two version of React Router (3 and 4) which had very different APIs. Therefore, version 2.x breaks redux-auth-wrapper into many more small, compostable pieces. redux-auth-wrapper still provides a simple import for those looking to get started quickly with React Router, but also allows for developers to import the building blocks to create redirection wrappers that work with any router (or history directly). You can even use redux-auth-wrapper 2.x in a project without redux or even routing.
```
Comment (piuccio) [inline docs/Migrating.md]: I think they are `composable pieces`, not `compostable pieces`

Comment (mjrussell) [inline docs/Migrating.md]: Thanks, I wrote this late last night so Im sure there are lots of typos 😉

**Comment by mjrussell** [comment]
Merged! Sorry for the delay. Theres a **ton** of new documentation (likely with some typos as @piuccio found already) on the new version. I also published an updated beta tag (2.0.0-beta3). Theres a few breaking changes from beta2, most notably:
1. `authenticatedSelector` instead of `predicate` and `authSelector`.
2. No more passing down `authData` which was the result of `authSelector`
3. No longer need to import `redux-auth-wrapper/lib/blah`, it should behave like the examples now

**Comment by mjrussell** [comment]
@piuccio you are right about the behavior of the predicate and `authenticatingSelector`, Im hoping that removing the predicate in favor of `authenticatedSelector` and `authenticatingSelector` clears it up a little. Although I probably need some extra documentation still.

As you can see from the [code](https://github.com/mjrussell/redux-auth-wrapper/blob/master/src/authWrapper.js#L36-L42), the result of `authenticatedSelector` is checked first, then the result of `authenticatingSelector`. I have a hard time deciding which of these should "win" and since the `predicate` use to take priority it seems to make sense to keep it in this order.

If you want the `authenticatingSelector` to take priority, you could have `authenticatedSelector` look at the same data that `authenticatingSelector` is:
```js
 authenticatedSelector: state => state.user.data !== null && !state.user.isLoading,
authenticatingSelector: state => state.user.isLoading
```

**Comment by piuccio** [comment]
wonderful 🎉 

If you publish the latest code on `beta3` or `rc1` I can try it out today or tomorrow

**Comment by mjrussell** [comment]
Great! 

The latest version should be pushed. It's 2.0-beta3

**Comment by mjrussell** [comment]
@piuccio 

> redirectPath is still mandatory when you use a FailureComponent even if it's never used.

This shouldn't be true with `connectedRouterRedirect` or `connectedReduxRedirect`. I removed `FailureComponent` from the API since its only really necessary with hiding or alternate components. Do you have an example?

Also if you've got a react native example, I'd love to add that to our getting started guide or code examples.

**Comment by piuccio** [comment]
I'm using this

```
// Library
import { connectedRouterRedirect } from 'redux-auth-wrapper/history4/redirect';
// Project
import {
  makeSelectUserName,
  makeSelectIsInitializingAuth,
} from 'containers/Auth/selectors';
import Loading from 'components/Loading';
import Login from 'containers/Login';

const initialising = makeSelectIsInitializingAuth();
const authSelector = makeSelectUserName();

/**
 * I've written this so many times and it always confuses me.
 *
 * authenticatingSelector: true, authenticatedSelector: true -> allow access
 * authenticatingSelector: true, authenticatedSelector: false -> allow access
 * authenticatingSelector: false, authenticatedSelector: false -> failure component
 * authenticatingSelector: false, authenticatedSelector: true -> loading component
 */
export const withAuth = connectedRouterRedirect({
  wrapperDisplayName: 'Authenticated',
  authenticatedSelector: (state) => (initialising(state) === false && !!authSelector(state)),
  authenticatingSelector: initialising,
  AuthenticatingComponent: Loading,
  FailureComponent: Login,
  redirectPath: '/',
});
```

I do have an example for react native, it's pretty much the same thing, I'm using the `FailureComponent` to render a login page. It's based on `beta1` or `beta2` can't remember.
When I find time to update to `beta3` or final I'll post it somewhere...

**Comment by mjrussell** [comment]
So in your code, you aren't actually redirecting anymore if you are displaying an alternative `FailureComponent`. This is because the default `FailureComponent` for a `connectedRouterRedirect` actually performs the redirect instead of mounting. You can see that [here](https://github.com/mjrussell/redux-auth-wrapper/blob/master/src/redirect.js). 

Is there a reason you don't want to have a route for `/login` and redirect users there? If you'd rather display one component vs the other instead of redirecting users with react router, then you should use the [Hiding and Alternate Components](https://mjrussell.github.io/redux-auth-wrapper/docs/Getting-Started/HidingAlternate.html). Your code would look like:

```js
import connectedAuthWrapper from 'redux-auth-wrapper/connectedAuthWrapper'

export const withAuth = connectedAuthWrapper({
  wrapperDisplayName: 'Authenticated',
  authenticatedSelector: (state) => (initialising(state) === false && !!authSelector(state)),
  authenticatingSelector: initialising,
  AuthenticatingComponent: Loading,
  FailureComponent: Login,
});

**Comment by piuccio** [comment]
In the other project (react-router 3 and auth-wrapper 1) I have an actual route for `/login` but because of the complexity of my authSelectors I often ended up having an infinite redirect loop between `/login` and `?redirectBack`, so in this other project I'm simply showing the `Login` as fallback.

I'll try your code, looks nicer.


In the meantime, this is what I'm using in react native + react-router 4

```js
export const Authenticated = connectedRouterRedirect({
  wrapperDisplayName: 'Authenticated',
  authSelector: makeSelectUserData(),
  authenticatingSelector: initialising,
  AuthenticatingComponent: LoadingPage,
  predicate: (user) => (user && user.type === 'whatever'),
  redirectPath: '/logout',
  allowRedirectBack: true,
});
```

the only trick to make it work is that you need to include these in your `package.json`

```
"redux-auth-wrapper": "2.0.0-beta",
"query-string": "4.3.4",
"url": "0.11.0"
```

**Pull request merged**